src/orca-jedi/interpolator/Interpolator.cc

```diff
@@ -177,6 +178,128 @@ template void Interpolator::executeInterpolation<float>(
+    atlas::field::MissingValue mv(inc.incrementFields()[gv_varname]);
+    bool has_mv = static_cast<bool>(mv);
+    for (std::size_t klev=0; klev < varSizes[jvar]; ++klev) {
+      for (std::size_t iloc=0; iloc < nlocs_; iloc++) {
+        if (has_mv && mv(field_view(iloc, klev))) {
+          result[out_idx] = util::missingValue<double>();
+        } else {
+          result[out_idx] = field_view(iloc, klev);
+        }
+        ++out_idx;
+      }
+    }
+  }
+}
+
+/// \brief Interpolate from observation space to model space
+/// \param vars Oops variables
+/// \param inc Increment object (output)
+/// \param mask Mask (observation space) vector
+/// \param resultin Values (observation space) vector (input)
+void Interpolator::applyAD(const oops::Variables& vars, Increment& inc,
+             const std::vector<bool> & mask,
+             const std::vector<double> & resultin) const
+{
+  // input is resultin output is inc
+
+  oops::Log::trace() << "orcamodel::Interpolator::applyAD start "
+                     << std::endl;
+
+  const size_t nvars = vars.size();
+
+  for (size_t j=0; j < nvars; ++j) {
+    if (!inc.variables().has(vars[j])) {
+      std::stringstream err_stream;
+      err_stream << "orcamodel::Interpolator::apply varname \" "
+                 << "\" " << vars[j]
+                 << " not found in the model increment." << std::endl;
+      err_stream << "    add the variable to the increment variables and "
+                 << "add a mapping from the geometry to that variable."
+                 << std::endl;
+      throw eckit::BadParameter(err_stream.str(), Here());
+    }
+  }
+
+  const std::vector<size_t> varSizes =
+    inc.geometry()->variableSizes(vars);
+  size_t nvals = 0;
+
+  for (size_t jvar=0; jvar < nvars; ++jvar) nvals += nlocs_ * varSizes[jvar];
+
+  std::size_t out_idx = 0;
+  for (size_t jvar=0; jvar < nvars; ++jvar) {
+    auto gv_varname = vars[jvar].name();
+    auto tgt_field = atlasObsFuncSpace_.createField<double>(
+      atlas::option::name(gv_varname) |
+      atlas::option::levels(varSizes[jvar]));
+
+// Copying observation array vector to an atlas observation field (tgt_field)
+    auto field_view = atlas::array::make_view<double, 2>(tgt_field);
+    atlas::field::MissingValue mv(inc.incrementFields()[gv_varname]);
+    bool has_mv = static_cast<bool>(mv);
+
+    for (std::size_t klev=0; klev < varSizes[jvar]; ++klev) {
+      for (std::size_t iloc=0; iloc < nlocs_; iloc++) {
+          if (has_mv && mv(field_view(iloc, klev))) {
+             field_view(iloc, klev) = util::missingValue<double>();
+          } else {
+             field_view(iloc, klev) = resultin[out_idx];
+          }
+        ++out_idx;
+      }
+    }
+
+// halo exchange update ghost points
+    std::shared_ptr<const Geometry> geom = inc.geometry();
+    geom->functionSpace().haloExchange(inc.incrementFields()[gv_varname]);
+
+    interpolator_.execute_adjoint(inc.incrementFields()[gv_varname], tgt_field);
```

Comment:
I might be wrong, but would it be good to apply a mask of the missing values at this point? I am exploring collaborating with SOCA and others in OOPS so that our land-sea-mask-aware interpolations work in a consistent and performant way. I am not suggesting doing anything in this PR, just exploring options!

Reply:
I'm not sure. I think I was expecting that interpolator_.execute_adjoint would (have to) be mask aware. Currently the increments are not masked anyway as they are not initialised from reading in a file (instead using the zero method). I could add a code comment here to this effect to remind me to look at this again later.

I certainly don't want to block any SOCA collaboration! Travis Sluka in particular was very helpful in getting me stated with JEDI DA. And it would be very cool to compare the NEMOVAR and SOCA DA results in orca-jedi!  

I'm happy to work on this more in future PRs and work with you on any adaptations needed in future to the code I'm putting into orca-jedi to facilitate any SOCA collaborations.